Treeview widget

codetent opened the pull request:
Implementation of treeview widget (#4)

Akuli commented:
This is nice :D I read through the code quickly, and it mostly looks good. I need to read and review this in more detail later.

Why do you need error handling in `FallbackConfigDict`? Does that give errors when a `TreeviewColumn` or `TreeviewRow` hasn't been added to a `Treeview` yet?

codetent commented:
The main idea behind the `FallbackConfigDict` is that when a tcl command cannot be executed, a temporary value should be set or read which is synchronized afterwards. 

I added this functionality because `TreeviewRow` and `TreeviewColumn` are useless before they are added to a `Treeview` (limitation of tk). If the object is not added, the `_treeview` attribute is not set of the class and therefore an exception is thrown which leads to accessing the fallback values.

I caught the generic expression for keeping the `FallbackConfigDict` usable for other purposes where another exception (not AttributeError) is raised.

PurpleMyst commented on docs/treeview.rst:
```diff
@@ -0,0 +1,61 @@
+The *rows* and *columns* attributes of the :class:`.Treeview` behave like normal
+ lists and therefore, items can be inserted and deleted as with lists. The only 
+exceptions are considering the direct setting of rows (*__setitem__*) and the 
+manipulation of the first column which is both prohibited. 
```
This should be "which are both prohibited"

codetent replied:
Oh, that's right. Thanks!

PurpleMyst commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,452 @@
+
+            for handler in handlers:
+                handler(None, option, value)
+        except Exception:
```
Why are you using `except Exception` here? What kinds of exceptions are you guarding against?

codetent replied:
> I caught the generic expression for keeping the `FallbackConfigDict` usable for other purposes where another exception (not AttributeError) is raised.

In this case, an `AttributeError` is raised but the fallback value should also be returned if another exception is thrown. It is not the best solution to catch a generic exception but for keeping it usable for other code which raises another kind of exception.

codetent replied:
Resolved by replacing the `FallbackConfigDict`.

PurpleMyst commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,452 @@
+                handler = handler[0]
+
+            return handler(self._types.get(option, str), option)
+        except Exception:
```
Same as above, which kinds of Exceptions are we guarding against here?

codetent replied:
Same answer as above, but problem is solved by change mentioned in the comment below.

PurpleMyst commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,452 @@
+    @make_thread_safe
+    def __init__(self, name=None, command=None, **kwargs):
+        self._name = name or 'C%d' % TreeviewColumn.next_col_num
+        self._treeview = kwargs.pop('treeview', None)
```
If you want `treeview` to be a keyword-only argument, it might be more correct to utilize the asterisk (`*`) "operator", as you can see in this example:

```python
def foo(a, *, b=None):
    return a * (b if b is not None else 0)

foo(1, b=2)  # returns 2
foo(1, 2)  # raises TypeError: foo() takes 1 positional argument but 2 were given
```

codetent replied:
Ah ok, I will change that! Thanks.

PurpleMyst commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,452 @@
+        return self._data[index]
+
+    def __setitem__(self, index, row):
+        raise NotImplementedError('use insert() instead')
```
What's wrong with implementing this using `TreeviewRowList.insert` and `TreeviewRowList.delete`?

codetent replied:
I implemented the management of rows and columns like a list. Therefore, `.insert()` is available anyway. For deleting items, `del item` can be used. I think this is more pythonic than an additional `.delete()` method.

Akuli commented:
> I added this functionality because TreeviewRow and TreeviewColumn are useless before they are added to a Treeview

There is a similar situation with the `Canvas` widget, and I solved it by just not allowing the users to create canvas items before they create the canvas, and it works well. I'm not perfectly happy with the canvas code. I think it would be better to not create an Item subclass like `type('Item', (CanvasItem,), {'canvas': self})` for each `Canvas` instance, and instead just pass the canvas as an argument whenever a `CanvasItem` object is created.

There is also a similar situation with the `Menu` widget, but it's solved differenly there, because menu items are used like this

```python3
window.config['menu'] = teek.Menu([
    teek.MenuItem("File", [
        teek.MenuItem("New", hello),
        teek.MenuItem("Open", hello),
        teek.MenuItem("Save", hello),
        teek.MenuItem("Quit", hello),
    ]),
    teek.MenuItem("Edit", [
        teek.MenuItem("Cut", hello),
        teek.MenuItem("Copy", hello),
        teek.MenuItem("Paste", hello),
    ]),
])
```

If the menu would need to be created first, then it would look like this

```python3
file_menu = teek.Menu()
file_menu.add_item("New", hello)
file_menu.add_item("Open", hello)
file_menu.add_item("Save", hello)
file_menu.add_item("Quit", hello)

edit_menu = teek.Menu()
edit_menu.add_item("Cut", hello)
edit_menu.add_item("Copy", hello)
edit_menu.add_item("Paste", hello)

window.config['menu'] = teek.Menu()
window.config['menu'].add_submenu(file_menu)
window.config['menu'].add_submenu(edit_menu)
```

which is much more annoying. This doesn't happen with canvas items, because canvas items can't be put inside other canvas items, like menus can be added to other menus as submenus.

codetent commented:
I thought about implementing this like the menu widget, but I was not sure if this is really the best way for the user.

I think I will change this behaviour by removing the `FallbackDict` and raise an exception if the user tries to set options of a row or column which is not added to a `Treeview`.

codetent commented:
The last changes replace the `FallbackDict` with a dictionary which allows multiple handler methods. Additionally, I simplified the code and added more tests.

There was also a problem with the tests using the newest svglib library. I created an issue for that problem!

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,469 @@
+        Insert column or column text converted to a column at given index.
+        Note that the column at index 0 cannot be replaced!
+        """
+        if index == 0:
```
This check isn't very good, because e.g. `-100` will likely insert at the beginning

```
>>> asdf = [1, 2, 3]
>>> asdf.insert(-100, 'lol')
>>> asdf
['lol', 1, 2, 3]
>>> 
```

something like this instead (I didn't try to run it):

```
if index < -len(self):
    index = -len(self)
if index < 0:
    index += len(self)
if index == 0:
    # this runs when inserting to the beginning
```

or even better, let the list handle all this, and check `if self.data[0] == column` after the insert.

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,469 @@
+                         reverse=reverse)
+
+        for old, row in enumerate(self.rows):
+            self.rows.move(old, sorted_rows.index(row))
```
Sort is a good idea, and I have seen tkinter users who have wanted it in tkinter.

This code feels wrong to me for a couple reasons:

- You are modifying `self.rows` in a loop that loops over `self.rows`. That is generally a bad idea. For example, what does this code print?

    ```
    stuff = ['a', 'b', 'c', 'd', 'e']
    for thing in stuff:
        stuff.remove(thing)
    print(stuff)
    ```

    The problem is that Python uses an index variable for looping internally, so it does `stuff.remove(stuff[0])`, `stuff.remove(stuff[1])` etc until `stuff[the index]` gives `IndexError`. The list shrinks as items are removed from it, which screws up the indexes; first this removes `a` (element 0), and then because the list is `['b', 'c', 'd', 'e']` it removes `'c'` (element 1), and then the list is `['b', 'd', 'e']` and it removes `'e'` (element 2). A simple solution is to loop over `stuff.copy()`.

- There is a similar issue with `self.rows.move`. Once an element has been moved, its index may still change as other items are moved. For example, if the rows are `[A, B, C]` after moving row `B` to index 1, and row `A` is moved to index 2, then `B` will be at index 0.

Here is a simpler example code that does not sort a list correctly:

```python3
import random

stuff = [7, 3, 2, 5, 6, 1, 8, 4]

def move(old, new):
    value = stuff[old]
    del stuff[old]
    stuff.insert(new, value)

sorted_stuff = sorted(stuff)
for old, value in enumerate(stuff):
    move(old, sorted_stuff.index(value))

print(stuff)
```

I would try something like this

```python3
for row in reversed(sorted_rows):
    self.rows.move(self.rows.index(row), 0)
```

This moves each item to the beginning, moving all the previously moved items after it. This is not ideal either, and it will create a *lot* of moves. There is probably a better way to do this, but at least the `move` method should be implemented with Tk's `move` command (search for `pathname move` in ttk_treeview manual page).

I think it's fine if you delete `sort` from this pull request and let me or someone else add it :)

Akuli replied:
This alternative seems to also work, and it would create less unnecessary moves especially when the treeview is already almost sorted (creates 0 moves if the treeview is already sorted):

```python3
for index, row in enumerate(sorted_rows):
    if self.rows[index] != row:
        self.rows.move(self.rows.index(row, index), index)
```

codetent commented:
Finally, I added the changes from the review! Now, nesting of rows is also possible and headings can be configured independently from columns.

codetent commented:
The CI tests on versions 3.4 and nightly fails now, because lxml does not support python 3.4 anymore and flake8 has a problem with the latest nightly build.

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,545 @@
+        if column_pos < 1:
+            raise IndexError('column position must be greater or equal 1')
+
+        sorted_rows = list(self).copy()
```
It's enough to do `list(self)`, no need to copy that newly created list.

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,545 @@
+            raise IndexError('column position must be greater or equal 1')
+
+        sorted_rows = list(self).copy()
+        sorted_rows.sort(key=lambda r: r.config['values'][column_pos - 1],
```
Why minus 1? Most programmers are familiar with things that start at zero, so why to make this start at 1 instead?

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,545 @@
+                         reverse=reverse)
+
+        for to_pos, row in enumerate(sorted_rows):
+            from_pos = self.index(row)
```
This will be quite slow with a big treeview, because finding the index of a row is slow and you're doing it in a loop. (If you know some big-O notation, this code is `O(n^2)` while sorting is usually `O(n*log(n))`.) Maybe mention the possible slowness in the docstring?

Alternatively, you might be able to avoid the whole `self.index(row)` by just deleting the `if from_pos` check?

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,545 @@
+        Similar to the ``config`` attribute that widgets have. The available
+        options are documented as ``ITEM OPTIONS`` in :man:`ttk_treeview(3tk)`.
+    """
+    _next_row_num = 0
```
Should this really be global to the whole program instead of being specific to a treeview instance?

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,545 @@
+        self._treeview._call(None, self._treeview, 'selection', 'add', self)
+
+    @make_thread_safe
+    def deselect(self):
```
When you name methods differently than the corresponding command in the manual page, can you mention that in the docstring? Something like "this calls `selection remove` documented in blah blah". You could also name this Python method `selection_remove`.

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,545 @@
+
+    @make_thread_safe
+    def __delitem__(self, index):
+        if index == 0:
```
This won't detect `del columnlist[0:1]` and `del columnlist[len(treeview) - 1]`. The slice thing is quite difficult to check, so you could just raise an error if someone attempts to delete a slice of the column list (and mention in the docs that deleting a slice of a column list isn't supported). It's easy to check for the `len-1` thing.

Akuli commented on teek/_widgets/treeview.py:
```diff
@@ -0,0 +1,545 @@
+
+    @make_thread_safe
+    def __setitem__(self, index, column):
+        if index == 0 or index <= -len(self):
```
This check will prevent setting the last column with `columnlist[-1] = something`. Better to use `==` instead of `<=`. There's also a similar thing in `insert`.
